[ADD] estate: create new module

utpat-odoo opened the pull request:
1. estate module 
Create new module estate module for manage estate property with feature like create new property manage buy and sell system 
2. Awesome OWL  
Create Card, Counter and Todo component 
3. Awesome Dashboard 
add customers and leads button in layout,add number card and pie card in dashboard, add configuration for card add and remove

utpat-odoo committed:
[ADD] estate: create new module

create new module estate and create model for new module and add security

utpat-odoo committed:
[ADD] estate: create views and security folder

Create module Real Estate
add menu list
set column default, read-only

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1016.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1016)

bit-odoo commented on estate/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# empty
```
You can remove this. It is unnecessary changes.

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,10 @@
+        'views/estate_property_views.xml',
+        'views/estate_menus.xml'
+        ]
+}
```
There should be one empty line at the end of the file.

bit-odoo commented on estate/models/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import estate_property
```
There should be one empty line at the end of the file.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,9 @@
+        <field name="view_mode">list,form</field>
+        <field name="context">{"search_default_available":1}</field>
+    </record>
+</odoo>
```
There should be one empty line at the end of the file.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,23 @@
+from odoo import models, fields
+from dateutil.relativedelta import relativedelta
+
+class EstateProperty(models.Model):
```
Leave two lines between import and classes for better readability.

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,10 @@
+    'name':'Estate',
+    'depends': ['base'],
+    'license': 'AGPL-3',
+    'data':[
```
```suggestion
    'data': [
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,23 @@
+    garage = fields.Boolean()
+    garden = fields.Boolean()
+    garden_area = fields.Integer()
+    garden_orientation=fields.Selection(string='Garden Orientation',selection=[('north', 'North'), ('south', 'South'), ('east','East'), ('west','West')])
```
```suggestion
    garden_orientation=fields.Selection(string='Garden Orientation',selection=[('north', 'North'), ('south', 'South'), ('east', 'East'), ('west', 'West')])
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,23 @@
+    garden_area = fields.Integer()
+    garden_orientation=fields.Selection(string='Garden Orientation',selection=[('north', 'North'), ('south', 'South'), ('east','East'), ('west','West')])
+    active=fields.Boolean(default=False)
+    state=fields.Selection(selection=[('new','New'),('offer_received','Offer Received'),('offer_accepted','Offer Accepted'),('sold','Sold'),('cancelled','Cancelled')])
```
```suggestion
    state=fields.Selection(
        selection=[
            ('new','New'),
            ('offer_received','Offer Received'),
            ('offer_accepted','Offer Accepted'),
            ('sold','Sold'),
            ('cancelled','Cancelled')
        ]
    )
```

bit-odoo reviewed (commented):
Hello, Good Work!
I have left some comments.

When you push your new code, please do not merge it with the previous commit.
 And make sure that your runbot is always green.
 
 Thanks

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+<odoo>
+    <record id="test_model_action" model="ir.actions.act_window">
```
You can use a proper name here like `estate_property_action`

utpat-odoo committed:
[IMP] estate: add list view and form view

In estate module add basic views list view and form view
Also add filter for new or offer received stage

utpat-odoo committed:
[IMP] estate: add list view and form view and improve syntax

In estate module add basic views list view and form view
Also add filter for new or offer received stage
improve syntax in estate_property model

bit-odoo commented on estate/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from . import models 
+
+
```
unnecessary diff.

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,10 @@
+{
+    'name':'Estate',
```
```suggestion
    'name': 'Estate',
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,39 @@
+        selection=[('north', 'North'),
+        ('south', 'South'), 
+        ('east','East'), 
+        ('west','West')
```
```suggestion
        ('east', 'East'), 
        ('west', 'West')
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,39 @@
+            ('offer_received','Offer Received'),
+            ('offer_accepted','Offer Accepted'),
+            ('sold','Sold'),
+            ('cancelled','Cancelled')
```
```suggestion
            ('new', 'New'),
            ('offer_received', 'Offer Received'),
            ('offer_accepted', 'Offer Accepted'),
            ('sold', 'Sold'),
            ('cancelled', 'Cancelled')
```

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,79 @@
+                <field name="facades"/>
+                <filter string="Available" name="group_by_available" domain="['|' ,('state', '=', 'new'), ('state', '=', 'offer_received')]"/>
+            </search>
+
```
unnecessary diff.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,79 @@
+        </field>
+    </record>
+
+
```
unnecessary diff.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,79 @@
+        </field>
+    </record>
+
+
```
unnecessary diff.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,79 @@
+        </field>
+    </record>
+
+
```
unnecessary diff.

bit-odoo reviewed (commented):
Hello, Good work!

I have left some remarks.
As I mentioned in my last comment, make sure that your runbot is always green.

utpat-odoo committed:
[IMP] estate: add property type,tag, offers, buyer and salesperson

Add new menu item property type and tags
Add buyer and salesperson fields
Add offer section

utpat-odoo committed:
[IMP] estate: add compute and on change field and add object type action

compute field like total area best offer
add validate and deadline fields with computation
add object type action on sold and cancel button and buyer field
add accept and refuse button in offer form

bit-odoo commented:
Hello!
Can you please make sure that your runbot is green?
https://runbot.odoo.com/runbot/tutorials-12?search=-utpat

Please update your PR title and message.

utpat-odoo committed:
[IMP] estate: add constraints in model

add SQL and Python both type constraints in database

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,95 @@
+from odoo import models, fields, api
+from dateutil.relativedelta import relativedelta
+from odoo.exceptions import UserError, ValidationError
+from odoo.tools.float_utils import float_compare
```
```suggestion
from dateutil.relativedelta import relativedelta

from odoo.exceptions import UserError, ValidationError
from odoo import models, fields, api
from odoo.tools.float_utils import float_compare
```

First external libraries and then internal import.

utpat-odoo committed:
[IMP] estate: add sprinkles

add status bar in property
set order in model
add handle list view of property type
add tag color

utpat-odoo committed:
[IMP] estate: add stat button, default filter, decorations

Implemented stat button on property type
default filter in property page
decorations in property offer and property
make editable list view for property tag

utpat-odoo committed:
[IMP] estate: add python, model, view inheritance

restrict user to delete property which is not in new or sold state
add property id field in user model using inheritance

utpat-odoo committed:
[IMP] estate: create new module estate account

created a new module to connect Real Estate with accounting features.
create invoice when property sold

utpat-odoo committed:
[IMP] estate: add kanban view

add kanban view
group by with property id
disable drag and drop in kanban view

utpat-odoo committed:
[IMP] estate: polish code coding guidelines wise

Improve syntax of menu item
Improve syntax in all views
Set id name  coding guidelines wise

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,105 @@
+            else:
+                record.state = 'sold'
+
+    _check_price = models.Constraint(
```
The constraint should be defined after the field declaration. You can take an example from Odoo code.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,105 @@
+    def _unlink_check_state_of_property(self):
+        for record in self:
+            if record.state in ('offer_received', 'offer_accepted', 'sold'):
+                raise UserError("You cannot delete a new or cancelled property !")
```
condition and error message not related:thinking:

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,66 @@
+    @api.model
+    def create(self, vals):
+        for record in vals:
+            property = self.env['estate.property'].browse(record['property_id'])
```
If possible, then we can move orm method outside the loop. You can get all the properties outside the loop and filter data inside the loop.

bit-odoo commented on estate/models/estate_property_tag.py:
```diff
@@ -0,0 +1,16 @@
+    _unique_name = models.Constraint(
+    'UNIQUE(name)',
+    'name already exists!',
+    )
```
```suggestion
    _check_unique_name = models.Constraint(
        'UNIQUE(name)',
        'name already exists!',
    )
```

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,139 @@
+                <field name="selling_price"/>
+                <field name="facades"/>
+                <field name="property_type_id"/>
+                <filter string="Available" name="available" domain="['|' ,('state', '=', 'new'), ('state', '=', 'offer_received')]"/>
```
```suggestion
                <filter string="Available" name="available" domain="[('state', 'in', ['new', 'offer_received'])]"/>
```

bit-odoo reviewed (commented):
Hello, i have added some comments please have a look into it.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,139 @@
+                        <group>
+                            <field name="tag_ids" widget="many2many_tags" options="{'color_field': 'color'}"/>
+                            <field name="property_type_id" options="{'no_create': true, 'no_open': true, 'no_edit': true}"/>
+                            <field name="name"/>
```
Name field should be displayed first for better visibility.

utpat-odoo committed:
[IMP] awesome_owl: add counter and card components

Make estate module UI user-friendly
Add counter and card component and display in awesome owl module

bit-odoo commented on awesome_owl/static/src/playground.xml:
```diff
@@ -1,10 +1,11 @@
+        <Card title="'This is title1'" content="'Hello from card1'"/>
+        <Card title="'This is title2'" content="'Hello from card2'"/>
     </t>
-
```
unnecessary diff.

utpat-odoo committed:
[IMP] awesome_owl: add todo list

add todo list in awesome_owl

utpat-odoo committed:
[IMP] awesome_dashboard: add customers and Leads button in layout

add controlpanel layout
add customers and leads button in controlpanel layout
add delete todo and toggle todo funtion
add property id field after customer field in account move form

bit-odoo commented on awesome_dashboard/static/src/dashboard.xml:
```diff
@@ -1,8 +1,15 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <templates xml:space="preserve">
-
```
unnecessary diff.

bit-odoo commented on awesome_dashboard/static/src/dashboard.xml:
```diff
@@ -1,8 +1,15 @@
+            </t>
+        </Layout>
     </t>
-
```
unnecessary diff.

bit-odoo commented on awesome_owl/static/src/card/card.xml:
```diff
@@ -0,0 +1,21 @@
+        <t t-if="this.state.visible">
+                <div class="card-body">
+                    <t t-slot="default"/>
+                </div>
```
Wrong indentation.

bit-odoo commented on awesome_owl/static/src/counter/counter.xml:
```diff
@@ -0,0 +1,37 @@
+        <p>Counter: <t t-esc="state.value"/></p>
+        <button class="increment-button" t-on-click="increment">
+            Increment
+          </button>
```
same.

bit-odoo commented on awesome_owl/static/src/playground.js:
```diff
@@ -1,5 +1,26 @@
+
+    }
+
+    totalSum(newVal,Checker){
```
```suggestion
    totalSum(newVal,Checker) {
```

bit-odoo commented on awesome_owl/static/src/playground.js:
```diff
@@ -1,5 +1,26 @@
+    }
+
+    totalSum(newVal,Checker){
+        if(Checker)
```
```suggestion
        if (Checker) {
```

bit-odoo commented on awesome_owl/static/src/playground.xml:
```diff
@@ -1,10 +1,23 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <templates xml:space="preserve">
-
```
unnecessary diff.

bit-odoo commented on awesome_owl/static/src/todo/todo_item.xml:
```diff
@@ -0,0 +1,11 @@
+<templates xml:space="preserve">
+    <t t-name="awesome_owl.TodoItem">
+        <div class="p-1" t-att-class="props.todo.isCompleted ? 'text-decoration-line-through text-muted' : ''">
+             <input class="form-check-input" type="checkbox" t-on-click="onChangeToggle"/>
```
indentation issue.

bit-odoo reviewed (commented):
Hello,
I have added some comments. Please do not repeat the same mistake again and again.

utpat-odoo committed:
[IMP] awesome_dashboard: add item list and pie chart in dashboard

add dashboard item which is number card and pie chart card
add lazy loading functionality for dashboard
making our dashboard generic
added functionality add and remove dashboard items
